operators/src/pro/ml/xgboost.rs

```diff
@@ -0,0 +1,560 @@
+        // given a set of raster bands, if they have different types,
+        // where should they be converted?
+        let vec_of_rqps: Vec<Box<dyn RasterQueryProcessor<RasterType = f64>>> = self
+            .sources
+            .iter()
+            .map(|init_raster| init_raster.query_processor().unwrap().into_f64())
+            .collect();
+
+        Ok(QueryProcessorOut(Box::new(XgboostProcessor::new(
+            vec_of_rqps,
+            self.model_file_path.clone(),
+        ))))
+    }
+}
+
+struct XgboostProcessor<Q, P>
+where
+    Q: RasterQueryProcessor<RasterType = P>,
+{
+    sources: Vec<Q>,
+    model_file: String,
+}
+
+impl<Q, P> XgboostProcessor<Q, P>
+where
+    Q: RasterQueryProcessor<RasterType = P>,
+    P: Pixel,
+{
+    pub fn new(sources: Vec<Q>, model_file_path: String) -> Self {
+        Self {
+            sources,
+            model_file: std::fs::read_to_string(model_file_path).unwrap(),
+        }
+    }
+
+    async fn process_tile_async(
+        &self,
+        bands_of_tile: Vec<Result<Tile<P>, crate::error::Error>>,
+        pool: Arc<ThreadPool>,
+    ) -> Result<RasterTile2D<PixelOut>> {
+        // gather information
+        let reference_tile = bands_of_tile.first().unwrap();
+        let tile = reference_tile.as_ref().unwrap().clone();
+        let n_rows = tile.grid_shape_array()[0];
+        let n_cols = tile.grid_shape_array()[1];
+        let grid_shape = tile.grid_shape();
+        let props = tile.properties;
+        let n_bands = bands_of_tile.len() as i32;
+        let time = tile.time;
+        let tile_position = tile.tile_position;
+        let global_geo_transform = tile.global_geo_transform;
+
+        // extract the actual data from the tiles
+        let rasters: Vec<Vec<f32>> = bands_of_tile
+            .into_iter()
+            .map(|band| {
+                let band_ok = band.unwrap();
+                let mat_tile = band_ok.into_materialized_tile();
+
+                mat_tile
+                    .grid_array
+                    .inner_grid
+                    .data
+                    .into_iter()
+                    .map(num_traits::AsPrimitive::as_)
+                    .collect::<Vec<f32>>()
+            })
+            .collect();
+
+        let mut pixels: Vec<f32> = Vec::new();
+
+        for row in 0..(n_rows * n_cols) {
+            let mut row_data: Vec<f32> = Vec::new();
+            for col in 0..n_bands {
+                let pxl = rasters
+                    .get(col as usize)
+                    .unwrap()
+                    .get(row)
+                    .unwrap()
+                    .to_owned();
+                row_data.push(pxl);
+            }
+            pixels.extend_from_slice(&row_data);
+        }
+
+        let model = self.model_file.clone();
+        let predicted_grid: geoengine_datatypes::raster::Grid<GridShape<[usize; 2]>, f32> =
+            crate::util::spawn_blocking(move || {
```

Comment:
I think you should do all the stuff above in spawn_blocking. You shouldn't do too much work in `async` functions

Reply:
Restructured in https://github.com/geo-engine/geoengine/pull/639/commits/5927879ce4601e18e31e87a860066f5d6794581d